docs/glossary.rst

```diff
@@ -298,13 +298,20 @@ Glossary
      foo` and `group bar`.
 
    userid
-     A *userid* is a string used to identify and authenticate
-     a real-world user or client. A userid is supplied to an
-     :term:`authentication policy` in order to discover the user's
-     :term:`principals <principal>`. In the authentication policies which
-     :app:`Pyramid` provides, the default behavior returns the user's userid as
-     a principal, but this is not strictly necessary in custom policies that
-     define their principals differently.
+     A *userid* is the string representation of an :term:`identity`.  Just like
+     the identity, it should identify the user associated with the current
+     request.  Oftentimes this is the ID of the user object in a database.
+
+   identity
+     An identity is an object identifying the user associated with the
+     current request.  The identity can be any object, but should implement a
+     ``__str__`` method that outputs a corresponding :term:`userid`.
```

Comment:
Hi!  Sorry to comment after this is merged, but I was following the issue and not aware of this PR.

I don’t understand the reason for using `__str__` as the method to convert Identity to user ID.
- `__str__` is meant to control what happens with `print`, `%s` formatting, etc.
- IMO it is good practice to have `__str__` fast, without side effects, without possibility of failure (so no parsing, casting, etc)
- `repr` uses `__str__` if the class doesn’t define `__repr__`
- REPL, logging, others use repr or str
- depending on the system, user ID may not be a string
- calling `str(identity)` instead of `identity.some_explicit_method()` seems a bit obscure (not discoverable)

If people agree with the disadvantages here, could this API decision be reconsidered?

Reply:
This is a discussion I'm happy to have. I'm not in love with `__str__` myself. I'd ask that since this is merged we open a new issue to talk about it though, instead of here.

Reply:
There are 2 good options that come to my mind: either return `storable_id, IdentityClass` tuple or have another function in the policy that derives the serializable id from the identity class...

Reply:
There's been some discussion on this, I dug up what I could find:

https://github.com/Pylons/pyramid/issues/3422#issuecomment-443840114
https://github.com/Pylons/pyramid/pull/3465#issuecomment-511140572

Reply:
Opened an issue over at #3520